### ISSUE TITLE
Redirector (`cmsd`) fails with lengthy tokens via HTTP (Internal error processing file)

### ISSUE THREAD

**Issue by olifre** [xrootd/xrootd#2247]
For a non-WLCG project, we are using lengthy tokens embedding groups and authorizations:
```
$ echo $BEARER_TOKEN | wc -c
3491
```
Accessing a DTN directly works:
```
$ curl -i -L -s --capath /etc/grid-security/certificates -H "Authorization: Bearer $BEARER_TOKEN" https://xrootd001.physik.uni-bonn.de:1094//cephfs/grid/helmholtz
HTTP/1.1 200 OK
Connection: Keep-Alive
Server: XrootD/v5.6.2
Content-Length: 3939
...
```
Using these to access a redirector fails, though:
```
$ curl -i -L -s --capath /etc/grid-security/certificates -H "Authorization: Bearer $BEARER_TOKEN" https://xrootd.physik.uni-bonn.de:1094//cephfs/grid/helmholtz
HTTP/1.1 500 Internal Server Error
Connection: Close
Server: XrootD/v5.6.2
Content-Length: 32

Internal error processing file.
```
Logs on the redirector show:
```
240424 09:40:43 22635 XrootdBridge: unknown.108250:37@myhost login as nobody
240424 09:40:43 22635 acc_Audit:  grant  *@? stat /cephfs/grid/helmholtz
240424 09:40:43 22635 cms_Pup: string too long packing char arg 5: opaque.
240424 09:40:43 22635 cms_Pack: Unable to pack request; too much data for code 10
240424 09:40:43 22635 XrootdXeq: unknown.108250:37@myhost disc 0:00:01 (send failure)
```
Without deep investigation, my guess is that the token is packed into the URL as `authz` (we are using `http.header2cgi Authorization authz`) and things get too long when passed to `cmsd`. Given the length of the used token, we are naturally exceeding the limit here:
https://github.com/xrootd/xrootd/blob/619e93fe896f23ac04fd424ac5223c2443fec296/src/XrdOuc/XrdOucPup.hh#L102

If such lengthy tokens should not work, feel free to close the issue (but I thought I'd open it for future reference and since differences in behaviour between DTN and when CMS lookup is performed via a Redirector may cause confusion in the future). 

As expected, the same lengthy token works via the XRootd protocol and ZTN without problems.

**Comment by abh3** [comment]
Thanks for pointing this out. Indeed, it would appear that the token plus any other CGI hat might be present is greater than 2K. The limit is arbitrary, we didn't think anyone would specify that much CGI information. It would be interesting to know what that token was trying to describe. It's likely it is listing every single path that the client is allowed to access. If so, people will quickly realize they have a lot of other issues. The solution here is to remove the authz element from the CGI as the cmsd has no need for it and it also removes one additional place where the token could be inadvertently exposed.

In order to set a priority for fixing this, is this something that you are encountering frequently and needs to be fixed sooner than later?

**Comment by olifre** [comment]
Many thanks for the quick reaction, as usual! 

This can certainly be put on low priority — we can workaround the problem for now by reducing the scopes embedded in the token. 

In the long run, though, lengthy tokens will be unavoidable with the AAI solution chosen in that project (choice mostly due to political reasons, not technical reasons :frowning_face: ). Currently, the token does not even contain path-based authorizations yet (but it should, in the future, which is likely quite some months away, though). 
To share my pain, here is a slightly sanitized token payload:
```json
{
  "entitlements": [
    "urn:mace:dir:entitlement:common-lib-terms",
    "urn:geant:dfn.de:nfdi.de:punch:group:PUNCH4NFDI:punch_intra#login.helmholtz.de",
    "urn:geant:dfn.de:nfdi.de:punch:group:PUNCH4NFDI:elsa_one:elsa_one_adm#login.helmholtz.de",
    "urn:geant:h-df.de:group:HDF#login.helmholtz.de",
    "urn:geant:dfn.de:nfdi.de:punch:group:PUNCH4NFDI#login.helmholtz.de",
    "urn:geant:dfn.de:nfdi.de:punch:group:PUNCH4NFDI:elsa_one#login.helmholtz.de"
  ],
  "sub": "35970c10-cc16-4490-9854-ccb14d9b7a30",
  "voperson_external_affiliation": "member@my-domain",
  "iss": "https://login.helmholtz.de/oauth2",
  "preferred_username": "user-id",
  "client_id": "public-oidc-agent",
  "eduperson_principal_name": "my-principal@domain.com",
  "eduperson_entitlement": [
    "urn:mace:dir:entitlement:common-lib-terms",
    "urn:geant:dfn.de:nfdi.de:punch:group:PUNCH4NFDI:punch_intra#login.helmholtz.de",
    "urn:geant:dfn.de:nfdi.de:punch:group:PUNCH4NFDI:elsa_one:elsa_one_adm#login.helmholtz.de",
    "urn:geant:h-df.de:group:HDF#login.helmholtz.de",
    "urn:geant:dfn.de:nfdi.de:punch:group:PUNCH4NFDI#login.helmholtz.de",
    "urn:geant:dfn.de:nfdi.de:punch:group:PUNCH4NFDI:elsa_one#login.helmholtz.de"
  ],
  "scope": "openid display_name sn single-logout offline_access voperson_id voperson_external_affiliation entitlements org_domain email profile credentials eduperson_scoped_affiliation eduperson_entitlement eduperson_principal_name eduperson_unique_id eduperson_assurance sys:scim:read_profile sys:scim:read_memberships sys:scim:read_self_group",
  "eduperson_scoped_affiliation": "affiliate@login.helmholtz.de",
  "eduperson_unique_id": "some-uuid-string@login.helmholtz.de",
  "sn": "Lastname",
  "exp": 1714015880,
  "iat": 1714011880,
  "jti": "...",
  "email": "my-mail@my-domain.de",
  "email_verified": true,
  "given_name": "Firstname",
  "display_name": "Fistname Lastname",
  "aud": "public-oidc-agent",
  "voperson_id": "some-uid@login.helmholtz.de",
  "name": "Firstname Lastname",
  "eduperson_assurance": [
    "https://refeds.org/assurance/IAP/medium",
    "https://refeds.org/assurance/IAP/local-enterprise",
    "https://refeds.org/assurance/ATP/ePA-1m",
    "https://refeds.org/assurance/ATP/ePA-1d",
    "https://refeds.org/assurance/IAP/low",
    "https://refeds.org/assurance",
    "https://aarc-project.eu/policy/authn-assurance/assam"
  ],
  "family_name": "Lastname"
}
```
Main issue are the very lengthy group names, listed in two claims, that `urn`-format is indeed based on an RFC for group names in tokens, but very lengthy. Since the AAI does not support claim filtering (yet), reduced sets of groups can not be requested (we have requested that feature, though...). Also, the `eduperson_assurance" claim needed for other services in the future eats up quite some characters. 

So in short, those tokens are indeed quite painful, but the AAI solution is not too uncommon (Unity IAM), so potentially there might be other projects using it at some point. We can workaround the size limitation for now by reducing scopes, but may run into it later on when path-based authorizations will be added. Thanks a lot again!

**Comment by abh3** [comment]
OK, then we will schedule a fix for R6 due some time in late Summer or early Fall.

**Comment by olifre** [comment]
Sadly, I only came around re-testing this with XRootD 6.0.3 now, the project was progressing a bit slower than expected. 

The unhappy news: I can still reproduce the problem — let me know in case you rather want a new issue or reopen this one:
```
$ echo $BEARER_TOKEN | wc -c
2052

$ curl -i -L -H "Authorization: Bearer $BEARER_TOKEN" -X GET --capath /etc/grid-security/certificates "https://xrootd-dev.physik.uni-bonn.de:1094/data/grid/helmholtz/"
HTTP/1.1 400 Bad Request
Connection: Close
Server: XRootD
Content-Length: 32

Internal error processing file.
```
This is with XRootD 6.0.3 (and of course matching CMSD), config contains `http.header2cgi Authorization authz` as usual. 

Part of debug logs are basically similar:
```
260603 12:18:52 706657 http_Protocol: getDataOneShot BuffAvailable: 1048576 maxread: 1048576
260603 12:18:52 706657 http_Protocol: getDataOneShot sslavail: 1048576
260603 12:18:52 706657 http_Protocol: read 2193 of 1048576 bytes
260603 12:18:52 706657 http_Protocol:  rc:36 got hdr line: GET /data/grid/helmholtz/ HTTP/1.1

260603 12:18:52 706657 http_Protocol:  Parsing first line: GET /data/grid/helmholtz/ HTTP/1.1

260603 12:18:52 706657 http_Protocol:  rc:42 got hdr line: Host: xrootd-dev.physik.uni-bonn.de:1094

260603 12:18:52 706657 http_Protocol:  rc:25 got hdr line: User-Agent: curl/8.19.0

260603 12:18:52 706657 http_Protocol:  rc:13 got hdr line: Accept: */*

260603 12:18:52 706657 http_Protocol:  rc:2075 got hdr line: Authorization: Bearer REDACTED

260603 12:18:52 706657 http_Protocol:  rc:2 got hdr line: 

260603 12:18:52 706657 http_Protocol:  rc:2 detected header end.
260603 12:18:52 706657 XrootdBridge: unknown.5:24@clientmachine login as nobody
260603 12:18:52 706657 unknown.5:24@clientmachine http_Protocol:  Process. lp:0x7f4218005918 reqstate: 0
260603 12:18:52 706657 unknown.5:24@clientmachine http_Protocol: Setting monitor info curl/8.19.0
260603 12:18:52 706657 unknown.5:24@clientmachine Xrootd_Protocol: 0000 Bridge req=3018 dlen=24 blen=24
260603 12:18:52 706657 unknown.5:24@clientmachine Xrootd_Protocol: 0000 set monitor info curl/8.19.0
260603 12:18:52 706657 unknown.5:24@clientmachine Xrootd_Response: 0000 sending 4 data bytes
260603 12:18:52 706657 http_Req:  XrdHttpReq::Data! final=False
260603 12:18:52 706657 unknown.5:24@clientmachine http_Req: PostProcessHTTPReq req: 2 reqstate: 0 final_:True
260603 12:18:52 706657 unknown.5:24@clientmachine http_Protocol:  Process. lp:(nil) reqstate: 0
260603 12:18:52 706657 unknown.5:24@clientmachine http_Req: Appended header fields to opaque info: 'authz=Bearer REDACTED'
260603 12:18:52 706657 unknown.5:24@clientmachine Xrootd_Protocol: 0000 Bridge req=3010 dlen=2089 blen=2089
260603 12:18:52 706657 unknown.5:24@clientmachine http_Protocol: Process is exiting rc:0
260603 12:18:52 706657 unknown.5:24@clientmachine Xrootd_Protocol: 0000 open rtS /data/grid/helmholtz/
260603 12:18:52 706657 unknown.5:24@clientmachine ofs_open: 10000000-600 (600) fn=/data/grid/helmholtz/
260603 12:18:52 706657 cms_Pup: string too long packing char arg 5: opaque.
260603 12:18:52 706657 cms_Pack: Unable to pack request; too much data for code 10
260603 12:18:52 706657 unknown.5:24@clientmachine Xrootd_Response: 0000 sending err 3000: Internal error processing file.
260603 12:18:52 706657 http_Req:  XrdHttpReq::Error
260603 12:18:52 706657 unknown.5:24@clientmachine http_Req: PostProcessHTTPReq req: 2 reqstate: 0 final_:False
260603 12:18:52 706657 unknown.5:24@clientmachine http_Protocol: Sending resp: 400 header len:83
```
(only client hostname replaced in log snippet)

**Comment by olifre** [comment]
My C++ is (sadly) getting barely used these days, but I tried my hand with `gdb` and could confirm:
* `XrdOucEnv::EnvTidy` is called just fine. 
* It triggers `XrdOucEnv::EnvBuildTidy` once. 
* This seems to correctly rebuild the environment without `authz` information, which in my example case means "empty env", stores that in `TIDY_ENVVAR`, and then restores the `global_env`. So the `authz` remains in the `global_env` and only the tidy env is cleaned. 
* `XrdOucEnv::EnvTidy` then appears to return the standard (uncleaned) env, which is too long for packing. 

So I think the problem is here:
https://github.com/amadio/xrootd/blob/ace1d2a5aa3d0e3081a558337a069bfbc57bf05b/src/XrdOuc/XrdOucEnv.cc#L153-L163
i.e. if the first "tidying" yields an empty environment, as it only contained the authz information, the env will not be cleaned. 

Following this assumption, I issued a request:
```
$ curl -i -L -H "Authorization: Bearer $BEARER_TOKEN" -X GET --capath /etc/grid-security/certificates "https://xrootd-dev.physik.uni-bonn.de:1094/data/grid/helmholtz/?foo"

HTTP/1.1 302 Found
Connection: Keep-Alive
Server: XRootD
Content-Length: 0
Location: https://NODE_TO_WHICH_REDIRECTION_HAPPENS:1094/data/grid/helmholtz/?authz=_very_long_token_

HTTP/1.1 403 Forbidden
Connection: Close
Server: XRootD
Content-Length: 56

Unable to open /data/grid/helmholtz/; permission denied
```
This is expected for this example case. 

So the fix seems not to work if the env purely contains `&authz=thebigtoken`. 

Adding at least one useless parameter to the request works around the problem.

**Comment by amadio** [comment]
Could you please create a new ticket for this? Otherwise the milestone assignments will be mixed up. We can link the new issue to this one. Cheers,

**Comment by olifre** [comment]
> Could you please create a new ticket for this? Otherwise the milestone assignments will be mixed up. We can link the new issue to this one. Cheers,

Sure, done and linked 👍 .
